ql/src/java/org/apache/hadoop/hive/ql/optimizer/OperatorGraph.java

```diff
@@ -0,0 +1,231 @@
+        dagIdx = min;
+        for (Edge<V, E> e : edges) {
+          if(e.t == this) {
+            DagNode<V, E> s = (DagNode<V, E>) e.s;
+            s.ensureDagIdxAtLeast(min + 1);
+          }
+        }
+      }
+    }
+
+    @Override
+    protected Node<V, E> newNode(V s) {
+      return new DagNode<V, E>(s);
+    }
+  }
+
+  DagGraph<Operator<?>, OpEdge> g;
+
+  enum EdgeType {
+    FLOW, SEMIJOIN, DPP, TEST,
+  }
+
+  static class OpEdge {
+
+    private final EdgeType et;
+    private final int index;
+
+    public OpEdge(EdgeType et) {
+      this(et, 0);
+    }
+
+    public OpEdge(EdgeType et, int index) {
+      this.et = et;
+      this.index = index;
+    }
+
+  }
+
+
+  Map<Operator<?>, Cluster> nodeCluster = new HashMap<>();
+
+  public class Cluster {
+
+    Set<Operator<?>> members = new LinkedHashSet<>();
+
+    public void merge(Cluster o) {
+      for (Operator<?> node : o.members) {
+        add(node);
+      }
+      o.members.clear();
+    }
+
+    public void add(Operator<?> curr) {
+      nodeCluster.put(curr, this);
+      members.add(curr);
+    }
+
+  }
+
+
+  public OperatorGraph(ParseContext pctx) {
+    g = new DagGraph<Operator<?>, OperatorGraph.OpEdge>();
+    Set<Operator<?>> visited = Sets.newIdentityHashSet();
+    Set<Operator<?>> seen = Sets.newIdentityHashSet();
+
+    seen.addAll(pctx.getTopOps().values());
+    while (!seen.isEmpty()) {
+      Operator<?> curr = seen.iterator().next();
+      seen.remove(curr);
+      if (visited.contains(curr)) {
+        continue;
+      }
+
+      visited.add(curr);
+
+      Cluster currentCluster = nodeCluster.get(curr);
+      if (currentCluster == null) {
+        currentCluster=new Cluster();
+        currentCluster.add(curr);
+      }
+      List<Operator<?>> parents = curr.getParentOperators();
+      for (int i = 0; i < parents.size(); i++) {
+        Operator<?> p = parents.get(i);
+        g.putEdgeValue(p, curr, new OpEdge(EdgeType.FLOW, i));
+        if (p instanceof ReduceSinkOperator) {
+          // ignore cluster of parent RS
+          continue;
+        }
+        Cluster cluster = nodeCluster.get(p);
+        if (cluster != null) {
+          currentCluster.merge(cluster);
+        } else {
+          currentCluster.add(p);
+        }
+      }
+
+      SemiJoinBranchInfo sji = pctx.getRsToSemiJoinBranchInfo().get(curr);
+      if (sji != null) {
+        g.putEdgeValue(curr, sji.getTsOp(), new OpEdge(EdgeType.SEMIJOIN));
+        seen.add(sji.getTsOp());
+      }
+      if (curr instanceof AppMasterEventOperator) {
+        DynamicPruningEventDesc dped = (DynamicPruningEventDesc) curr.getConf();
+        TableScanOperator ts = dped.getTableScan();
+        g.putEdgeValue(curr, ts, new OpEdge(EdgeType.DPP));
+        seen.add(ts);
+      }
+
+      List<Operator<?>> ccc = curr.getChildOperators();
+      for (Operator<?> operator : ccc) {
+        seen.add(operator);
+      }
+    }
+  }
+
+  public void toDot(File outFile) throws Exception {
```

Comment:
this is not called from anywhere right now - but I've used it during debug to get a better understanding of the plan

Reply:
yes; I've moved these classes - I use these methods during development; and they are very usefull; I needed to see the join tree - and I was able to extract a simple tree after a few minutes.

I plan to expose this kind of information on the hs2 web interface in some way.